core/src/main/java/org/apache/iceberg/encryption/EnvelopeEncryptionManager.java

```diff
@@ -0,0 +1,139 @@
+
+/**
+ * Encryption manager which in conjunction with a KMS can encrypt {@link OutputFile} and decrypt
+ * {@link InputFile}. Envelope encryption uses a key wrapping strategy, where a Key Encryption Key (KEK) is used to
+ * wrap or unwrap a Data Encryption Key DEK which is used to encrypt the underlying files.
+ *
+ * When generating new DEKs for OutputFiles, this class will first attempt to have the KMS generate a new key. If the
+ * KMS does not support key generation a new key will be produced by pulling bytes from a {@link SecureRandom} on the
+ * JVM writing the file.
+ */
+public class EnvelopeEncryptionManager implements EncryptionManager {
+  private final EnvelopeConfiguration dataEncryptionConfig;
+  private final boolean nativeFormatEncryption;
+  private final KmsClient kmsClient;
+  private final int dataKeyLength;
+
+  private transient volatile SecureRandom workerRNG = null;
+
+  /**
+   * @param nativeFormatEncryption true for native encryption inside file formats (Parquet or ORC), false for generic
+   *                               encryption of file streams
+   * @param dataEncryptionConfig configuration for encryption of data files. Future versions will add configuration for
+   *                             encryption of metadata files
+   * @param kmsClient Client of KMS used to wrap/unwrap keys in envelope encryption
+   * @param dataKeyLength If DEKs are not generated by KMS, they will be randomly generated by Iceberg using this length
+   */
+  public EnvelopeEncryptionManager(boolean nativeFormatEncryption,  EnvelopeConfiguration dataEncryptionConfig,
+      KmsClient kmsClient, int dataKeyLength) {
+    this.nativeFormatEncryption = nativeFormatEncryption;
+    if (!nativeFormatEncryption) {
+      throw new UnsupportedOperationException("EnvelopeEncryptionManager currently only supports encryption " +
+          "provided by the underlying file format.");
+    }
+    Preconditions.checkNotNull(dataEncryptionConfig,
+        "Cannot create EnvelopeEncryptionManager because data encryption config is not passed");
+    this.dataEncryptionConfig = dataEncryptionConfig;
+    Preconditions.checkNotNull(dataEncryptionConfig.kekId(),
+        "Cannot create EnvelopeEncryptionManager because table key encryption key ID is not specified");
+    Preconditions.checkNotNull(kmsClient,
+        "Cannot create EnvelopeEncryptionManager because KmsClient is null");
+    this.kmsClient = kmsClient;
+    this.dataKeyLength = dataKeyLength;
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    EnvelopeMetadata metadata = generateEnvelopeMetadata(dataEncryptionConfig);
+
+    if (nativeFormatEncryption) {
+      NativeFileCryptoParameters nativeEncryptParams = NativeFileCryptoParameters.create(metadata.dek())
+          .encryptionAlgorithm(metadata.algorithm())
+          .build();
+
+      Preconditions.checkArgument(rawOutput instanceof NativelyEncryptedFile,
```

Comment:
Not sure if the error message will do string concat even the condition is true. Otherwise, the concat could take some time and generate garbages because encrypt is called pretty often.

Reply:
a good point. This is called per data file, so not very often; but still, I think, a reason enough to revert this back to an if block.